osquery/utils/fgets_buffer.h

```diff
@@ -0,0 +1,201 @@
+/**
+ *  Copyright (c) 2019-present, osquery Foundation
+ *  All rights reserved.
+ *
+ *  This source code is licensed in accordance with the terms specified in
+ *  the LICENSE file found in the root directory of this source tree.
+ */
+
+#pragma once
+
+#include <cstdint>
+#include <memory>
+#include <string.h>
+#include <string>
+#include <vector>
+
+#include <osquery/filesystem/fileops.h>
+
+/*
+ * This all exists to ensure that we don't get hangs when
+ * using fgets() on pipes.  See issue #4810 for details.
+ */
+
+/**
+ * @brief Abstraction of a non blocking file for use with FgetsBuffer
+ */
+struct NonblockingFile {
+  virtual bool isValid() = 0;
+
+  virtual bool isDataAvail() = 0;
+
+  virtual int64_t read(std::vector<char>& buf) = 0;
+
+  virtual void close() = 0;
+
+  virtual ~NonblockingFile() {}
+};
+
+typedef std::unique_ptr<NonblockingFile> SPNonblockingFile;
```

Comment:
nitpick, we've been using `NonblockingFileRef` similar to Apple's conventions; prefer that or writing out the whole type.